Clean up upgrade.php

NanoSector opened the pull request:
When running upgrade.php I noticed a lot of issues pop up. I'm going to clean it up as much as I can to hopefully make maintenance easier.

NanoSector committed:
Remove unused global variable

Signed-off-by: Yoshi2889 <rick.2889@gmail.com>

NanoSector committed:
Fix undefined variable $substep

Signed-off-by: Yoshi2889 <rick.2889@gmail.com>

NanoSector committed:
Remove deprecated mysql_ functions in favor of mysqli_

Signed-off-by: Yoshi2889 <rick.2889@gmail.com>

NanoSector committed:
Fix clean_cache duplication; fixes #3755

Signed-off-by: Yoshi2889 <rick.2889@gmail.com>

NanoSector commented:
Do we really need CleanupMods? It seems to be touching a whole lot of stuff the upgrader shouldn't really be doing (it parses a modification's XML file to see if it's still valid, something the package manager in the admin panel does just fine). If you've just upgraded you really shouldn't be touching old mods.

NanoSector committed:
Remove unused and broken cleanupMods code

Signed-off-by: Yoshi2889 <rick.2889@gmail.com>

NanoSector committed:
Oops... re-disable $disable_security

Signed-off-by: Yoshi2889 <rick.2889@gmail.com>

NanoSector committed:
Remove unused upgrade_templates() template

Signed-off-by: Yoshi2889 <rick.2889@gmail.com>

NanoSector committed:
Add missing docblock for MySQLConvertOldIp

Signed-off-by: Yoshi2889 <rick.2889@gmail.com>

NanoSector committed:
Fix duplicate array keys

Signed-off-by: Yoshi2889 <rick.2889@gmail.com>

NanoSector committed:
Cleanup unused variables

Signed-off-by: Yoshi2889 <rick.2889@gmail.com>

albertlast commented:
When your already touching the upgrade.php,
an "step" who is missing the config rewrite from mysql to mysqli see the part of autoswitch: https://github.com/SimpleMachines/SMF2.1/issues/3583

NanoSector committed:
Set $db_type to mysqli during upgrade, if coming from mysql

Signed-off-by: Yoshi2889 <rick.2889@gmail.com>

jdarwood007 commented:
CleanupMods was added in 2.0 beta phase as a way to attempt to automatically reinstall mods.  It caused a lot of problems, so it was just disabled until we could get around to resolving those issues.  It never made it by the time 2.0 was gold.

NanoSector commented:
We probably should look into recreating it in a sane way, then. A lot has changed and we should be able to use the Subs files' functions now instead of manually checking packages.

sbulen commented:
Thanks Yoshi...   While in Upgrade.php, you might want to take a peek at issue #3554

NanoSector commented:
I'll take a look, thanks.

NanoSector committed:
Add global $txt so language files load properly...oops

Signed-off-by: Yoshi2889 <rick.2889@gmail.com>

NanoSector commented:
Why do we populate the entire user array during upgrade? Shouldn't it be enough to just check if the user has admin access?

NanoSector committed:
Fix progress bar during serialization to JSON

Signed-off-by: Yoshi2889 <rick.2889@gmail.com>

NanoSector committed:
Fix undefined index error on command line

Signed-off-by: Yoshi2889 <rick.2889@gmail.com>

NanoSector committed:
No need to check login if $disable_security is true

I'm not sure if we actually use the $upcontext['user'] array somewhere, other than to check for admin access.
Signed-off-by: Yoshi2889 <rick.2889@gmail.com>

NanoSector committed:
Fix ConvertUtf8 loop logic, should fix #3554

Signed-off-by: Yoshi2889 <rick.2889@gmail.com>

live627 commented:
The upgrader is still a complete mystery to me. Many thanks for fixing it.

NanoSector commented:
I kind of want to avoid a last-minute rewrite to further stretch the next release but I really feel that a new upgrade/installation framework should be put in place. It is a big mess indeed, lots of old code.

The thing that worries me most about the current upgrader is that there is a lot of code copied from the Sources files. IMO, this is a direct consequence of having huge functions in many of those files. For instance, checkLogin copies huge chunks of password checking etc. from LogInOut.php, but it cannot use the functions inside LogInOut.php because they were written for a specific use case (and can't be used for any other use case). This in the end means that if we ever decide to change the behavior of those functions (or even worse, fix a security issue), Upgrade.php will not reflect these changes, possibly leaving vulnerabilities open. Fixing this then gets messy because some chunks are copied directly from the Sources files, some are altered a bit, so we're essentially forced to fix the same thing twice.

Fixing this mess is possible by splitting code like LogInOut.php into smaller, specialized functions; Upgrade.php can then simply directly include LogInOut.php and utilize these functions. This avoids code duplication as the same code can be reused internally too and allows easier maintenance of code (and as a bonus gets us a much higher code rating in Scrutinizer).

I'm not sure if this should be done in 2.1 though as it pretty much means rewriting half the codebase, but it is something I would love to see in 3.0.

Oldiesmann commented:
At this point I think it's best to just clean up upgrade.php as much as we can and look at rewriting everything for the next version of SMF. It would be a lot of work to completely rewrite all of that at this point, especially given how long 2.1 has been in development.

NanoSector commented:
Yeah. I'll probably get some more cleaning up done today.

NanoSector committed:
Default to mysqli backend instead of mysql

Signed-off-by: Yoshi2889 <rick.2889@gmail.com>

NanoSector committed:
Only one driver type for mysql now

Signed-off-by: John Rayes <live627@gmail.com>

NanoSector committed:
Move helper functions inside upgrade-helper.php

Signed-off-by: Yoshi2889 <rick.2889@gmail.com>

NanoSector committed:
Docblocks for variables

Signed-off-by: Yoshi2889 <rick.2889@gmail.com>

NanoSector committed:
Remove legacy code

Signed-off-by: Yoshi2889 <rick.2889@gmail.com>

NanoSector committed:
Remove upgrade-helper.php when removing files

Signed-off-by: Yoshi2889 <rick.2889@gmail.com>

NanoSector commented:
When digging through the code I noticed we only recommend to backup your tables when upgrading from YaBB SE (and I changed that to always recommend it). That got me thinking though, do we really want to have the upgrade script backup tables? I mean, it's nice to have it, but shouldn't we encourage the user to make a full backup of their database first, since that's better than just copying tables?

NanoSector committed:
Always recommend the user to backup their database tables.

Signed-off-by: Yoshi2889 <rick.2889@gmail.com>

NanoSector committed:
Remove more mysql references in favor of mysqli

Signed-off-by: Yoshi2889 <rick.2889@gmail.com>

NanoSector committed:
Moar docblocks

Signed-off-by: Yoshi2889 <rick.2889@gmail.com>

NanoSector committed:
mb_strtolower exists in all php 5 and 7 versions; simplify smf_strtolower

Signed-off-by: Yoshi2889 <rick.2889@gmail.com>

NanoSector committed:
quickFileWritable now returns a boolean on every code path

Signed-off-by: Yoshi2889 <rick.2889@gmail.com>

NanoSector committed:
[WIP] Remove YaBB support, we have a dedicated converter now.

Signed-off-by: Yoshi2889 <rick.2889@gmail.com>

NanoSector committed:
Shove more helper functions in upgrade-helper.php

Signed-off-by: Yoshi2889 <rick.2889@gmail.com>

NanoSector committed:
textfield_alter: Remove MySQL 4.1 support

Signed-off-by: Yoshi2889 <rick.2889@gmail.com>

NanoSector committed:
textfield_alter: Add docblock

Signed-off-by: Yoshi2889 <rick.2889@gmail.com>

tinoest commented on other/upgrade.php:
```diff
@@ -28,15 +35,6 @@
-	),
 	'postgresql' => array(
 		'name' => 'PostgreSQL',
 		'version' => '8.0',
```
From this comment here; http://www.simplemachines.org/community/index.php?topic=547286.msg3890058#msg3890058

PostgreSQL is meant to be version 9.1 minimum now.

jdarwood007 commented:
Yabb != Yabb SE.
We still support upgrading from Yabb SE (see the upgrade_1.0.sql).  Yabb needs a converter.


Have we tested a SMF 1.1 or 2.0 install upgrading to 2.1 with mysqli?  Does the upgrader handler that smoothly?

NanoSector commented:
But we have converters for both of them:
http://www.simplemachines.org/community/index.php?topic=140776.0

I would test but I only have php7 on my systems, that doesn't work with
smf2.0 or 1.1.

Op wo 4 jan. 2017 01:45 schreef Jeremy D <notifications@github.com>:

> Yabb != Yabb SE.
> We still support upgrading from Yabb SE (see the upgrade_1.0.sql).
>
> Have we tested a SMF 1.1 or 2.0 install upgrading to 2.1 with mysqli? Does
> the upgrader handler that smoothly?
>
> —
> You are receiving this because you authored the thread.
> Reply to this email directly, view it on GitHub
> <https://github.com/SimpleMachines/SMF2.1/pull/3761#issuecomment-270267151>,
> or mute the thread
> <https://github.com/notifications/unsubscribe-auth/ABOJfG_ffVnWVFqJSJGl5Qvc9aDlDXoYks5rOuu8gaJpZM4LXzm4>
> .
>

NanoSector commented:
Sorry, didn't notice your comment; I'll fix that tomorrow.

Op di 3 jan. 2017 21:28 schreef Martyn Brown <notifications@github.com>:

> *@tinoest* commented on this pull request.
> ------------------------------
>
> In other/upgrade.php
> <https://github.com/SimpleMachines/SMF2.1/pull/3761#pullrequestreview-15005552>
> :
>
> > @@ -28,15 +35,6 @@
>  		'utf8_version_check' => 'global $db_connection; return mysqli_get_server_info($db_connection);',
>  		'alter_support' => true,
>  	),
> -	'mysql' => array(
> -		'name' => 'MySQL',
> -		'version' => '5.0.3',
> -		'version_check' => 'return min(mysql_get_server_info(), mysql_get_client_info());',
> -		'utf8_support' => true,
> -		'utf8_version' => '5.0.3',
> -		'utf8_version_check' => 'return mysql_get_server_info();',
> -		'alter_support' => true,
> -	),
>  	'postgresql' => array(
>  		'name' => 'PostgreSQL',
>  		'version' => '8.0',
>
> From this comment here;
> http://www.simplemachines.org/community/index.php?topic=547286.msg3890058#msg3890058
>
> PostgreSQL is meant to be version 9.1 minimum now.
>
> —
> You are receiving this because you authored the thread.
> Reply to this email directly, view it on GitHub
> <https://github.com/SimpleMachines/SMF2.1/pull/3761#pullrequestreview-15005552>,
> or mute the thread
> <https://github.com/notifications/unsubscribe-auth/ABOJfCfXHL8pyo2vlwSvO5jDdG-zVQdVks5rOq9JgaJpZM4LXzm4>
> .
>

live627 committed:
pg min ver is 9.1

Signed-off-by: John Rayes <live627@gmail.com>

live627 commented:
@Yoshi2889 do you wish to add more changes to this? Can I merge this?

sbulen commented:
I'm looking forward to this set of changes!

NanoSector commented:
I don't think I'd want to add more changes, thanks for fixing that postgre
issue. I cannot test the whole thing since I only have access to PHP 7 now
so if someone can test if it still upgrades fine from both 1.1 and 2.0 I
think it's good to go.

Op zo 8 jan. 2017 06:11 schreef sbulen <notifications@github.com>:

> I'm looking forward to this set of changes!
>
> —
> You are receiving this because you were mentioned.
> Reply to this email directly, view it on GitHub
> <https://github.com/SimpleMachines/SMF2.1/pull/3761#issuecomment-271130549>,
> or mute the thread
> <https://github.com/notifications/unsubscribe-auth/ABOJfCM8w41bnfthy_1JCnK5kZ0G9v-Mks5rQHAEgaJpZM4LXzm4>
> .
>

live627 committed:
Found a few functions to recycle + random bug fixes

Signed-off-by: John Rayes <live627@gmail.com>

live627 committed:
always ensure settings backup is writable since we always use it

Signed-off-by: John Rayes <live627@gmail.com>

live627 committed:
add proxy to upgrade

Signed-off-by: John Rayes <live627@gmail.com>

live627 committed:
use anon function for errror handler

Signed-off-by: John Rayes <live627@gmail.com>

Oldiesmann commented on other/install.php:
```diff
@@ -523,9 +523,8 @@ function CheckFilesWritable()
 		'Themes',
 		'agreement.txt',
 		'Settings.php'
```
You're missing a comma at the end of line 525

live627 committed:
that was my evil clone

NanoSector commented:
Didnt we change the mysqli think back to mysql? I think Upgrade.php has not been changed accordingly, then..

live627 commented:
I already changed it.

NanoSector commented:
Ah, cool.

live627 committed:
Some slightly random things I found while testing

Signed-off-by: John Rayes <live627@gmail.com>
